cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -185,5 +185,14 @@ inline bool operator==(const SubtreeImpl::Encoded& l, const SubtreeImpl::Encoded
          l.partition_expression == r.partition_expression;
 }
 
+template <typename T>
+std::shared_ptr<T> DowncastFragmentScanOptions(
+    const std::shared_ptr<ScanOptions>& scan_options, const std::string& type_name) {
+  if (!scan_options) return nullptr;
+  if (!scan_options->fragment_scan_options) return nullptr;
+  if (scan_options->fragment_scan_options->type_name() != type_name) return nullptr;
```

Comment:
I think this should be an error
```suggestion
  if (scan_options->fragment_scan_options->type_name() != type_name) {
    return Status::Invalid("FragmentScanOptions of type ", scan_options->fragment_scan_options->type_name(),
        " were provided for scanning a fragment of type ", type_name);
  }
```